cirq/circuits/circuit_operation.py

```diff
@@ -152,6 +152,28 @@ def _parameter_names_(self) -> AbstractSet[str]:
             )
         }
 
+    def mapped_circuit(self, deep: bool = False) -> 'cirq.Circuit':
+        circuit = self.circuit.unfreeze()
+        if deep:
+            circuit = circuit.map_ops(
+                lambda op: op.mapped_circuit(deep=True) if isinstance(op, CircuitOperation) else op
+            )
+        circuit = circuit.transform_qubits(lambda q: self.qubit_map.get(q, q))
+        circuit = protocols.with_measurement_key_mapping(circuit, self.measurement_key_map)
+        circuit = protocols.resolve_parameters(circuit, self.param_resolver, recursive=False)
+        if self.repetitions < 0:
+            circuit = circuit ** -1
+        if self.repetition_ids is None or not protocols.is_measurement(circuit):
+            return circuit * abs(self.repetitions)
+        keys = protocols.measurement_keys(circuit)
+        return cirq.Circuit(
+            protocols.with_measurement_key_mapping(circuit, {key: f'{rep}-{key}' for key in keys})
+            for rep in self.repetition_ids
+        )
+
```

Comment:
Just want to preemptively chime in on the draft PR. I think for `deep == False`, you'd still want to propagate `repetition_ids` for `CircuitOperations` and only map measurement_keys for direct child measurements like in `_decompose_` below. This is to avoid the pitfalls described in [repeated measurement RFC](http://docs/document/d/1_dY92lTKumaxH4wRdyabqQz_-i_xupNVwMzbrz__1U8?resourcekey=0-bcAZJNPDvnmthf5Y8VFVZQ#heading=h.gbik77iezhj5).

Also, it might be nice to have `mapped_op_tree` method that does what this does without converting to `Circuit`. That way, we can reuse this code for `mapped_circuit` and `_decompose_` both.

Reply:
+1 to reviewing pitfalls for `repetition_ids`, but I don't think `mapped_op_tree` is strictly necessary. If everything works as intended, `_decompose_` should just reduce to
```
def _decompose_(self):
    return self.mapped_circuit(False).all_operations()
```
`mapped_circuit` works better as the base method since `OP_TREE` doesn't preserve moments.